procrastinate-org/procrastinate
# Rework migrations

Closes #<ticket number>

WIP work, but I've been blocking for too long already on this.
This is a new take on acceptance tests:
- They're run as subprocess, as before
- But in a dedicated venv
- And can be run with multiple configuration (but the same tests will run)

As a start, I created 4 tests:
- Normal conditions (current code, all migrations)
- Run on the last release of procrastinate with just the pre-migrations (currently failing! @medihack, see below)
- Run on the current code with just the pre-migrations
- Normal conditions but use aiopg (currently not working, due to not specifying that we need the `aiopg` extra in the venv)

What's missing:
- Fix the 2 failing test
- Add some other combinations ?
- The operations we run on each test are very minimal, we should test a few more things (see existing acceptance tests)
- Remove acceptance tests once we test (mostly?) everything in the new tests. and remove this ugly "2" in the folder name
- Add LATEST_TAG envvar in the CI to avoid reading the latest tag from the git tags during the tests. I don't trust our CI git clone to do it properly. Instead use `gh releases something`
- Add a Readme explaining what the hell this whole thing is and how to contribute


### Successful PR Checklist:


#### PR label(s): 
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20breaking%20%F0%9F%92%A5
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20feature%20%E2%AD%90%EF%B8%8F
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20bugfix%20%F0%9F%95%B5%EF%B8%8F
  - [x] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20miscellaneous%20%F0%9F%91%BE
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20dependencies%20%F0%9F%A4%96
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20documentation%20%F0%9F%93%9A

The interesting error we got on running procrastinate 2.x with pre-migrations:
```
E           tests.acceptance2.utils.SubprocessException: Subprocess failed (code 1) with stdout:  and stderr: Launching a job: app.t1(a=1)
E           function procrastinate_defer_job_v1(unknown, unknown, smallint, unknown, unknown, jsonb, unknown) does not exist
E           LINE 1: SELECT procrastinate_defer_job_v1($1, $2, $3, $4, $5, $6, $7...
E                          ^
E           HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
E           Database error.
```
